Comment text object

habamax opened the pull request:
Add comment text object `ic` and `ac`:

[![asciicast](https://asciinema.org/a/akUDDNUUVf44rlX7QVptv3EOA.svg)](https://asciinema.org/a/akUDDNUUVf44rlX7QVptv3EOA)

Works when syntax is enabled, useful to do something with a comment (including uncommenting a comment with `gcic` or `gcac`)

habamax committed:
Add comment text object

habamax committed:
Correct information about toggling.

habamax committed:
add tests for comment text object

chrisbra commented on runtime/pack/dist/opt/comment/autoload/comment.vim:
```diff
@@ -76,3 +76,84 @@ export def Toggle(...args: list<string>): string
+export def ObjComment(inner: bool)
+    def IsComment(): bool
+        var stx = map(synstack(line('.'), col('.')), 'synIDattr(v:val, "name")')->join()
+        if stx =~ 'Comment'
```
```suggestion
        if stx =~? 'Comment'
```

chrisbra commented on runtime/pack/dist/opt/comment/autoload/comment.vim:
```diff
@@ -76,3 +76,84 @@ export def Toggle(...args: list<string>): string
+            return false
+        endif
+    enddef
+
```
```suggestion
    # requires syntax support
    if !exists("g:syntax_on")
      return
    endif
```

chrisbra commented on runtime/pack/dist/opt/comment/autoload/comment.vim:
```diff
@@ -76,3 +76,84 @@ export def Toggle(...args: list<string>): string
+
+    # Search for the beginning of the comment block
+    if IsComment()
+        if search('\v(\S+)|$', 'bW', 0, 200, IsComment) > 0
```
```suggestion
        if search('\v%(\S+)|$', 'bW', 0, 200, IsComment) > 0
```

chrisbra commented on runtime/pack/dist/opt/comment/autoload/comment.vim:
```diff
@@ -76,3 +76,84 @@ export def Toggle(...args: list<string>): string
+    # Search for the beginning of the comment block
+    if IsComment()
+        if search('\v(\S+)|$', 'bW', 0, 200, IsComment) > 0
+            search('\v(\S+)|$', 'W', 0, 200, () => !IsComment())
```
same here

chrisbra commented on runtime/pack/dist/opt/comment/autoload/comment.vim:
```diff
@@ -76,3 +76,84 @@ export def Toggle(...args: list<string>): string
+            search('\v(\S+)|$', 'W', 0, 200, () => !IsComment())
+        else
+            cursor(1, 1)
+            search('\S\+', 'cW', 0, 200)
```
here you are not using `\v`. Can we stay consistent with the use of `\v`?

habamax replied:
sure, although, I have used `\v` to simplify regexes where I thought it makes sense.

chrisbra commented on runtime/pack/dist/opt/comment/autoload/comment.vim:
```diff
@@ -76,3 +76,84 @@ export def Toggle(...args: list<string>): string
+        if search('\v(\S+)|$', 'bW', 0, 200, IsComment) > 0
+            search('\v(\S+)|$', 'W', 0, 200, () => !IsComment())
+        else
+            cursor(1, 1)
```
When would this else trigger? I am wondering when we would need to start jumping to the beginning of the file and start searching again.

habamax replied:
I will add a test case for this

https://asciinema.org/a/WTLvMeBAM1R6Ld7106iTFa0tb

chrisbra commented on runtime/pack/dist/opt/comment/autoload/comment.vim:
```diff
@@ -76,3 +76,84 @@ export def Toggle(...args: list<string>): string
+    var pos_start = getcurpos()
+
+    if !inner
+        if search('\s*', 'bW', line('.'), 200) > 0
```
That matches very often, even zero times. Can we make this `\s` instead or do we need `\s\+`?

habamax replied:
right `\s\+`

habamax replied:
Nah, we can't use search here, will be using while loop instead, example case:

![image](https://github.com/user-attachments/assets/12ecef11-6310-463f-9dd2-3e73e21b1fd1)

`vac` on `main` should select a comment, instead it selects a space and `{`:

```
int main() {// main start
}  /* main end */
```

habamax replied:
Added test case as well.

habamax replied:
```
    if !inner
        var col = pos_start[2]
        var prefix = getline(pos_start[1])[ : col - 2]
        while col > 0 && prefix[col - 2] =~ '\s'
            col -= 1
        endwhile
        pos_start[2] = col
        # instead of this:
        # if search('\v\s+', 'bW', line('.'), 200) > 0
        #     pos_start = getcurpos()
        # endif
    endif
```

chrisbra commented on runtime/pack/dist/opt/comment/doc/comment.txt:
```diff
@@ -32,10 +32,19 @@ back to `gcc` otherwise, add the following mapping to your vimrc: >
 Note: using `gC` may not always result in valid comment markers depending on
 the language used.
 
+Additionally, the plugin defines comment text-object which has effect when
```
```suggestion
Additionally, the plugin defines a comment text-object which requires syntax highlighting to be enabled.
```

habamax committed:
use =~? comparisson

habamax committed:
add syntax_on guard

habamax committed:
use \v in regexes

habamax committed:
update help

habamax committed:
add test for a comment text object on a first line

habamax committed:
use \v\s+ regex

habamax committed:
fix "no space before comment" issue

Should correctly select a comment without leading space:

print("hello");// this comment should work too

habamax committed:
add testcase for noleading space case

habamax committed:
fix regression, add test

habamax committed:
fix issue with inline comment

Should work with/*comments inside*/a string.

habamax committed:
Update Last update date

habamax committed:
add test cases for consequent comments

benknoble commented on runtime/pack/dist/opt/comment/autoload/comment.vim:
```diff
@@ -76,3 +76,92 @@ export def Toggle(...args: list<string>): string
+            return true
+        else
+            return false
+        endif
```
```suggestion
        return stx =~? 'Comment'
```

Perhaps?

habamax replied:
done

benknoble commented on runtime/pack/dist/opt/comment/doc/comment.txt:
```diff
@@ -32,10 +32,19 @@ back to `gcc` otherwise, add the following mapping to your vimrc: >
-depends on the first line of the range of lines to act upon.  When it matches
-a comment marker, the line will be un-commented, if it doesn't, the line will
+depends on the range of lines to act upon.  When all of the lines in range
+have comment markers, all lines will be un-commented, if it doesn't, the lines will
```
My eyeball suggests this line now overruns the margin, but I didn't measure precisely.

habamax replied:
eagle eye!

fixed.

habamax committed:
fix doc formatting

habamax committed:
simplify IsComment
